### ISSUE TITLE
[Feature request] Support if/else in evalNode

### ISSUE THREAD

**Issue by rydeng** [influxdata/kapacitor#741]
I want to use evalNode like this, 

```
stream
    |eval(lambda:"value" > 100? 30:40)
        .as("substitution")
```
It is just an example, I hope kapacitor's evalNode can have this feature. It's a greate promotion for evalNode and very useful.
^ - ^

**Comment by nathanielc** [comment]
Currently this can be done mathematically:

```
stream
    |eval(lambda: 10*(int("value" > 100) +3))
        .as("substitution")
```

That may not work in all cases but its a good workaround for now.

Explanation:

* int(boolean expression)  returns 0 or 1
* +3 becomes: 3 or 4
* * 10 becomes: 30 or 40

I do like your proposed syntax of the ternary operator, but it does represent a non trivial amount of work to add support for it.

**Comment by yosiat** [comment]
Hi @nathanielc ,
I would like to try to implement this, I think I should start with the lexing phase an mark "?" as "TokenTernaryStart" and ":" "TokenTernaryEnd".

Just to make sure my approach is valid: I need to add case for "?" in "lexOperator" and loop until I find ":" or EOF, something like this:
```
lastPos = l.pos
while(true) {
   t = l.next()
   if t == EOF {
      Error out Invalid tick
   }

   if t == ':' {
       Go back to lastPos, emit TokenTernaryStart
       Go back to ":" position, emit TokenTernaryEnd
   }
}
```

Is this ok? am In the right direction?

**Comment by yosiat** [comment]
@nathanielc I have thought of another two options to solve this:

1. Allow binary nodes to returns any primitive value, so this statement ``TRUE && 6`` will return 6 and this ``FALSE && 8`` will return false. then you can write ternary as something like - `` a ? b : c`` => ``(a && b) || c``
2. Add if function, like this - ``lambda: if(a, b, c)``

what do you think?

**Comment by nathanielc** [comment]
@yosiat Thanks for looking into this.

Before we look at implementation I want to make sure we do get the syntax right. Some initial thoughts/questions:

* Most important factor to me is that TICKscript remains readable.
* I like the ternary operator because its familiar and fits on a single line.
* TICKscript is a declarative language, once we  add conditional logic we start to enter the realm of an imperative language.  TICKscript is not meant to be a general purpose language. If you have the need for complex logic beyond what is provided by default you have two options. Use UDFs or add a new node type if the need is general enough.

OK, with all that said I am not against adding if/else kind of logic to TICKscript as it is generally useful, but I want to make sure we do so in an easy to read declarative way.


Lets examine the ternary expression. Here is a generic definition of the syntax.

`(boolean expressions) ? (expression) : (expression)`

Keep in mind that a  ternary expression is an expression also. So they can be nested, for example:

`"value" > 100 ? "value" > 150 ? 3 :  2 : 1`

Written another way:

```go
switch {
   case value > 150:
     return 3
   case value > 100:
     return 2
   default:
    return 1
}
```

Clearly the switch statement is easier to read than the nested ternary operators. But a switch statement is far beyond what we want and is in the realm of imperative language constructs.

Adding parenthesis helps its readability, but can still get out of hand quickly.

` "value" > 100 ? ("value" > 150 ? 3 :  2)  : 1`

I am open to suggestions/comments. We should nail down the syntax before we start implementation.

Have a look [here](https://en.wikipedia.org/wiki/%3F:) for some ideas.

As a side note: The reason we want to keep TICKscript as a declarative language is so that uses do not try to add significant amount of processing logic in their scripts. If we wanted TICKscript to be able to process the data directly in complex ways we would not have written a DSL and would have used a general purpose langue to describe Kapacitor tasks. This is how similar technologies like Apache Spark and Flink make use of Java.

**Comment by nathanielc** [comment]
@yosiat Just  saw you second comment:

As for option 1, I don't like it because then you have to mess with the type system and define truthy values for each of the primitive types. Every language that does this gets it wrong. Probably because there is no clean way to do it.

As for the second option it seems reasonable, evaluate it on what I posted above.

**Comment by yosiat** [comment]
@nathanielc We saw the need for some conditionals in tick script in our use case whether it was on evaluation or in tick script in templates (for example, if you have some parameter specified do align on window, otherwise don't).

Considering the complexity of this, I think it will be simpler to start with implementing "if" function, and then you can write - ``if(value > 150, 3, if(value > 100, 2, 1))``, since we evaluate the functions params as well.

What do you think?

**Comment by nathanielc** [comment]
Slight tangent...

>  (for example, if you have some parameter specified do align on window, otherwise don't).

For these cases we should make the `align` properties accept an optional boolean value, `align(should_align)`.

I like the `if` function as it doesn't require new operators or changes to the lexer/parser. It also reads well since it requires the parenthesis. Plus one from me.

**Comment by yosiat** [comment]
@nathanielc  ;)
I will submit PR tomorrow for the if function :)

About the tangent, this the same for alerts - we have ui, and user doesn't have to specify "info" / "warn" / "crit", so currently we put default value of "lambda: FALSE".. but this is not critical issue for us, if it will be I will create new separate issue for this.

**Comment by rydeng** [comment]
@nathanielc   @yosiat  you two are cool，thanks for reply.
